File: solutions/04.enterprise-bot/Dialogs/Cancel/CancelDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder.Dialogs;

namespace EnterpriseBot
{
    public class CancelDialog : ComponentDialog
    {
        // Constants
        public const string CancelPrompt = "cancelPrompt";

        // Fields
        private static CancelResponses _responder = new CancelResponses();

        public CancelDialog()
            : base(nameof(CancelDialog))
        {
            InitialDialogId = nameof(CancelDialog);

            var cancel = new WaterfallStep[]
            {
                AskToCancel,
                FinishCancelDialog,
            };

            AddDialog(new WaterfallDialog(InitialDialogId, cancel));
            AddDialog(new ConfirmPrompt(CancelPrompt));
        }

        public static async Task<DialogTurnResult> AskToCancel(WaterfallStepContext sc, CancellationToken cancellationToken) => await sc.PromptAsync(CancelPrompt, new PromptOptions()
        {
            Prompt = await _responder.RenderTemplate(sc.Context, "en", CancelResponses._confirmPrompt),
        });

        public static async Task<DialogTurnResult> FinishCancelDialog(WaterfallStepContext sc, CancellationToken cancellationToken) => await sc.EndDialogAsync((bool)sc.Result);

        protected override async Task<DialogTurnResult> EndComponentAsync(DialogContext outerDc, object result, CancellationToken cancellationToken)
        {
            var doCancel = (bool)result;

            if (doCancel)
            {
                // If user chose to cancel
                await _responder.ReplyWith(outerDc.Context, CancelResponses._cancelConfirmed);

                // Cancel all in outer stack of component i.e. the stack the component belongs to
                return await outerDc.CancelAllDialogsAsync();
            }
            else
            {
                // else if user chose not to cancel
                await _responder.ReplyWith(outerDc.Context, CancelResponses._cancelDenied);

                // End this component. Will trigger reprompt/resume on outer stack
                return await outerDc.EndDialogAsync();
            }
        }
    }
}

File: samples/javascript_typescript/52.enterprise-bot/src/dialogs/cancel/cancelDialog.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License

import { ComponentDialog, ConfirmPrompt, DialogContext, DialogTurnResult, WaterfallDialog, WaterfallStepContext } from "botbuilder-dialogs";
import { CancelResponses } from "./cancelResponses";

export class CancelDialog extends ComponentDialog {
    // Constants
    public static readonly CancelPrompt: string = "cancelPrompt";

    public static async AskToCancel(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        return await sc.prompt(CancelDialog.CancelPrompt, {
            prompt : await CancelDialog._responder.renderTemplate(sc.context, "en", CancelResponses._confirmPrompt),
        });
    }

    public static FinishCancelDialog(sc: WaterfallStepContext): Promise<DialogTurnResult> {
        return sc.endDialog(sc.result as boolean);
    }

    // Fields
    private static _responder: CancelResponses;

    constructor() {
        super(CancelDialog.name);
        this.initialDialogId = CancelDialog.name;

        const cancel = [
            CancelDialog.AskToCancel.bind(this),
            CancelDialog.FinishCancelDialog.bind(this),
        ];

        this.addDialog(new WaterfallDialog(this.initialDialogId, cancel));
        this.addDialog(new ConfirmPrompt(CancelDialog.CancelPrompt));
    }

    protected async endComponent(outerDC: DialogContext, result: any) {
        const doCancel: boolean = result;

        if (doCancel) {
            // If user chose to cancel
            await CancelDialog._responder.replyWith(outerDC.context, CancelResponses._cancelConfirmed);

            // Cancel all in outer stack of component i.e. the stack the component belongs to
            return await outerDC.cancelAllDialogs();
        } else {
            // else if user chose not to cancel
            await CancelDialog._responder.replyWith(outerDC.context, CancelResponses._cancelDenied);

            // End this component. Will trigger reprompt/resume on outer stack
            return await outerDC.endDialog();
        }
    }
}
